backend/internal/middleware/permission.go

```diff
@@ -0,0 +1,154 @@
+package middleware
+
+import (
+	"net/http"
+	"strconv"
+
+	"github.com/ChukwukaRosemary23/flowboard-backend/internal/services"
+	"github.com/gin-gonic/gin"
+)
+
+// RequirePermission middleware checks if user has specific permission on a board
+func RequirePermission(permissionName string) gin.HandlerFunc {
+	return func(c *gin.Context) {
+		// Get user ID from JWT (set by auth middleware)
+		userID, exists := c.Get("user_id")
+		if !exists {
+			c.JSON(http.StatusUnauthorized, gin.H{"error": "Unauthorized"})
+			c.Abort()
+			return
+		}
+
+		// Get board ID from URL parameter
+		boardIDStr := c.Param("board_id")
+		if boardIDStr == "" {
+			boardIDStr = c.Param("id") // For board routes like /boards/:id
+		}
+
+		boardID, err := strconv.ParseUint(boardIDStr, 10, 32)
+		if err != nil {
+			c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid board ID"})
+			c.Abort()
+			return
+		}
+
+		// Check permission
+		permService := &services.PermissionService{}
+		hasPermission := permService.CheckPermission(userID.(uint), uint(boardID), permissionName)
+
+		if !hasPermission {
+			c.JSON(http.StatusForbidden, gin.H{
+				"error": "You don't have permission to perform this action",
+			})
+			c.Abort()
+			return
+		}
+
+		c.Next()
+	}
+}
+
+// RequireOwner ensures only board owner can access
+func RequireOwner() gin.HandlerFunc {
+	return func(c *gin.Context) {
+		userID, exists := c.Get("user_id")
+		if !exists {
+			c.JSON(http.StatusUnauthorized, gin.H{"error": "Unauthorized"})
+			c.Abort()
+			return
+		}
+
+		boardIDStr := c.Param("id")
+		if boardIDStr == "" {
+			boardIDStr = c.Param("board_id")
+		}
+
+		boardID, err := strconv.ParseUint(boardIDStr, 10, 32)
+		if err != nil {
+			c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid board ID"})
+			c.Abort()
+			return
+		}
+
+		permService := &services.PermissionService{}
+		if !permService.IsOwner(userID.(uint), uint(boardID)) {
+			c.JSON(http.StatusForbidden, gin.H{
+				"error": "Only board owner can perform this action",
```

Comment:
When creating public APIs, you try not to be too explicit with some error messages for security reasons. However, you can log the actual error message/reason to console for audit reviews.

```suggestion
				"error": "You do not have permission to perform this action",
```